url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/classloading/CascadingClassLoadHelper.java

```diff
@@ -0,0 +1,310 @@
+   *
+   * @param name name of the desired resource
+   * @return a java.io.InputStream object
+   */
+  @Override
+  public InputStream getResourceAsStream(String name) {
+
+    InputStream result = null;
+
+    if (bestCandidate != null) {
+      result = bestCandidate.getResourceAsStream(name);
+      if (result == null) {
+        bestCandidate = null;
+      }
+    }
+
+    ClassLoadHelper loadHelper = null;
+
+    Iterator<ClassLoadHelper> iter = loadHelpers.iterator();
+    while (iter.hasNext()) {
+      loadHelper = iter.next();
+
+      result = loadHelper.getResourceAsStream(name);
+      if (result != null) {
+        break;
+      }
+    }
+
+    bestCandidate = loadHelper;
+    return result;
+  }
+
+  /**
+   * Given a package name, search the classpath for all classes that extend sundial.job
+   *
+   * @param pkgname
+   * @return
+   */
+  public Set<Class<? extends Job>> getJobClasses(String pkgname) {
+
+    Set<Class<? extends Job>> classes = new HashSet<Class<? extends Job>>();
+
+    String relPath = pkgname.replace('.', '/');
+
+    // Get a File object for the package
+    URL resource = getResource(relPath);
+    if (resource == null) {
+      throw new RuntimeException("Unexpected problem: No resource for " + relPath);
+    }
+    logger.info("Package: '" + pkgname + "' becomes Resource: '" + resource.toString() + "'");
+
+    if (resource.toString().startsWith("jar:")) {
+      processJarfile(resource, pkgname, classes);
+    } else {
+      processDirectory(new File(resource.getPath()), pkgname, classes);
+    }
+
+    return classes;
+
+  }
+
+  private void processDirectory(File directory, String pkgname, Set<Class<? extends Job>> classes) {
+
+    logger.debug("Reading Directory '" + directory + "'");
+    // Get the list of the files contained in the package
+    String[] files = directory.list();
+    for (int i = 0; i < files.length; i++) {
+      String fileName = files[i];
+      String className = null;
+      // we are only interested in .class files
+      if (fileName.endsWith(".class")) {
+        // removes the .class extension
+        className = pkgname + '.' + fileName.substring(0, fileName.length() - 6);
+      }
+      logger.debug("FileName '" + fileName + "'  =>  class '" + className + "'");
+      if (className != null) {
+        filterJobClassWithExceptionCatch(className, classes);
+      }
+      File subdir = new File(directory, fileName);
+      if (subdir.isDirectory()) {
+        processDirectory(subdir, pkgname + '.' + fileName, classes);
+      }
+    }
+  }
+
+  private void processJarfile(URL resource, String pkgname, Set<Class<? extends Job>> classes) {
+
+    String relPath = pkgname.replace('.', '/');
+    String resPath = resource.getPath().replace("%20", " ");
+    String jarPath = resPath.replaceFirst("[.]jar[!].*", ".jar").replaceFirst("file:", "");
+    logger.debug("Reading JAR file: '" + jarPath + "'");
+    JarFile jarFile;
+    try {
+      jarFile = new JarFile(jarPath);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *jarFile*. Execution paths that do not contain closure statements were detected. Ensure *jarFile* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).